transcom/mymove
# MAIN B-21592

[INT PR](https://github.com/transcom/mymove/pull/14029)
[INT PR #2](https://github.com/transcom/mymove/pull/14129)

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21592)

## Summary

There was an issue found where the wrong contract year was being selected for fees due to the date that it was considering was when the move was made available to Prime and not any dates where actual action was occurring. The gov asked us to use the requested pickup date on a shipment to be the driving factor for which contract year was selected for the move management & counseling fees.

This PR does the following:
- adds logic that pulls the first `requestedPickupDate` on a shipment within a move when creating a `MS` or `CS` service item
- adds and updates relevant tests

### How to test

1. Access MM as a customer, create a move (don't use devseed data - doesn't work for some reason)
2. Set up some shipments - not PPMs, make your first shipment's requested delivery date go back/forward a year or two
3. Go through SC flow, submit the move to TOO
4. Now view the move management fee on the MTO tab and it should pull that price from the `re_task_order_fees` table
5. Find that price and check out the `contract_year_id` column and traverse to the `re_contract_years` table
6. Confirm that the date of your requested pickup date falls between the `start_date` and `end_date`




## [Issue ticket](https://www13.v1host.com/USTRANSCOM38/Issue.mvc/Summary?oidToken=Issue%3A1029382)
## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21592)

## Summary

The man, myth, and legend @CoryKleinjanCACI found an issue following the initial merge of B-21592 that they were unable to approve full PPM shipments due to the lack of a `requestedPickupDate`. Since we want to ensure that the right base year is picked, I sent out a question to Dan Schuster and he confirmed that he wants to use a PPM's expected departure date value to decide the base year.

This PR does the following:
- updates query to use `Eager` when initially querying the move when creating a service item (`FetchOne` just wasn't cutting it since I couldn't use associations)
- added in 
```
	var nilTime time.Time
	if shipment.PPMShipment != nil && shipment.PPMShipment.ExpectedDepartureDate != nilTime {
		feeDate = &shipment.PPMShipment.ExpectedDepartureDate
	}
```
when creating a service item to cover down on PPM shipments
- added full PPM test

### How to test

1. Access MM as office/customer a create a move with full PPM (setting `NS Mayport` as current/destination)
2. Go in as the TOO, update/add orders information
3. Approve the PPM shipment and add the CS service item
4. Should be successful and should have an estimated price in the correct base year
5. Repeat with partial PPM
6. Should pull from NOT the PPM shipment (requested pickup date) for the correct base year


![Screenshot 2024-11-07 at 10 39 13 AM](https://github.com/user-attachments/assets/5a1fa7d6-4451-4bb0-bd4f-d9d90222cb0d)
